File: api/src/main/java/com/ruby/rubia_server/core/service/CompanyGroupService.java
package com.ruby.rubia_server.core.service;

import com.ruby.rubia_server.core.dto.CompanyGroupDTO;
import com.ruby.rubia_server.core.dto.CreateCompanyGroupDTO;
import com.ruby.rubia_server.core.dto.UpdateCompanyGroupDTO;
import com.ruby.rubia_server.core.entity.CompanyGroup;
import com.ruby.rubia_server.core.repository.CompanyGroupRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class CompanyGroupService {

    private final CompanyGroupRepository companyGroupRepository;

    public CompanyGroupDTO create(CreateCompanyGroupDTO createDTO) {
        log.info("Creating company group with name: {}", createDTO.getName());
        if (companyGroupRepository.findByName(createDTO.getName()).isPresent()) {
            throw new IllegalArgumentException("Company group with name '" + createDTO.getName() + "' already exists");
        }

        CompanyGroup companyGroup = CompanyGroup.builder()
                .name(createDTO.getName())
                .description(createDTO.getDescription())
                .isActive(createDTO.getIsActive())
                .build();

        CompanyGroup savedCompanyGroup = companyGroupRepository.save(companyGroup);
        log.info("Company group created successfully with id: {}", savedCompanyGroup.getId());
        return toDTO(savedCompanyGroup);
    }

    @Transactional(readOnly = true)
    public List<CompanyGroupDTO> findAll() {
        return companyGroupRepository.findAll().stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public CompanyGroupDTO findById(UUID id) {
        return companyGroupRepository.findById(id)
                .map(this::toDTO)
                .orElseThrow(() -> new IllegalArgumentException("Company group not found with id: " + id));
    }

    public CompanyGroupDTO update(UUID id, UpdateCompanyGroupDTO updateDTO) {
        log.info("Updating company group with id: {}", id);
        CompanyGroup existingCompanyGroup = companyGroupRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Company group not found with id: " + id));

        if (updateDTO.getName() != null && !updateDTO.getName().equals(existingCompanyGroup.getName())) {
            if (companyGroupRepository.findByName(updateDTO.getName()).isPresent()) {
                throw new IllegalArgumentException("Company group with name '" + updateDTO.getName() + "' already exists");
            }
            existingCompanyGroup.setName(updateDTO.getName());
        }

        if (updateDTO.getDescription() != null) {
            existingCompanyGroup.setDescription(updateDTO.getDescription());
        }

        if (updateDTO.getIsActive() != null) {
            existingCompanyGroup.setIsActive(updateDTO.getIsActive());
        }

        CompanyGroup updatedCompanyGroup = companyGroupRepository.save(existingCompanyGroup);
        log.info("Company group updated successfully with id: {}", updatedCompanyGroup.getId());
        return toDTO(updatedCompanyGroup);
    }

    public void delete(UUID id) {
        log.info("Deleting company group with id: {}", id);
        if (!companyGroupRepository.existsById(id)) {
            throw new IllegalArgumentException("Company group not found with id: " + id);
        }
        companyGroupRepository.deleteById(id);
        log.info("Company group deleted successfully");
    }

    private CompanyGroupDTO toDTO(CompanyGroup companyGroup) {
        return CompanyGroupDTO.builder()
                .id(companyGroup.getId())
                .name(companyGroup.getName())
                .description(companyGroup.getDescription())
                .isActive(companyGroup.getIsActive())
                .createdAt(companyGroup.getCreatedAt())
                .updatedAt(companyGroup.getUpdatedAt())
                .build();
    }
}

File: api-dotnet/Rubia.Server/Services/CompanyGroupService.cs
using Microsoft.EntityFrameworkCore;
using Rubia.Server.Data;
using Rubia.Server.DTOs;
using Rubia.Server.Entities;
using Rubia.Server.Services.Interfaces;

namespace Rubia.Server.Services;

public class CompanyGroupService : ICompanyGroupService
{
    private readonly RubiaDbContext _context;
    private readonly ILogger<CompanyGroupService> _logger;

    public CompanyGroupService(RubiaDbContext context, ILogger<CompanyGroupService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<CompanyGroupDto> CreateAsync(CreateCompanyGroupDto createDto)
    {
        _logger.LogInformation("Creating company group with name: {Name}", createDto.Name);

        if (await _context.CompanyGroups.AnyAsync(cg => cg.Name == createDto.Name))
        {
            throw new ArgumentException($"Company group with name '{createDto.Name}' already exists");
        }

        var companyGroup = new CompanyGroup
        {
            Name = createDto.Name,
            Description = createDto.Description,
            IsActive = createDto.IsActive
        };

        _context.CompanyGroups.Add(companyGroup);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Company group created successfully with id: {Id}", companyGroup.Id);
        return ToDto(companyGroup);
    }

    public async Task<List<CompanyGroupDto>> FindAllAsync()
    {
        var companyGroups = await _context.CompanyGroups
            .OrderBy(cg => cg.Name)
            .ToListAsync();

        return companyGroups.Select(ToDto).ToList();
    }

    public async Task<CompanyGroupDto> FindByIdAsync(Guid id)
    {
        var companyGroup = await _context.CompanyGroups
            .FirstOrDefaultAsync(cg => cg.Id == id);

        if (companyGroup == null)
        {
            throw new ArgumentException($"Company group not found with id: {id}");
        }

        return ToDto(companyGroup);
    }

    public async Task<CompanyGroupDto> UpdateAsync(Guid id, UpdateCompanyGroupDto updateDto)
    {
        _logger.LogInformation("Updating company group with id: {Id}", id);

        var existingCompanyGroup = await _context.CompanyGroups
            .FirstOrDefaultAsync(cg => cg.Id == id);

        if (existingCompanyGroup == null)
        {
            throw new ArgumentException($"Company group not found with id: {id}");
        }

        if (updateDto.Name != null && updateDto.Name != existingCompanyGroup.Name)
        {
            if (await _context.CompanyGroups.AnyAsync(cg => cg.Name == updateDto.Name))
            {
                throw new ArgumentException($"Company group with name '{updateDto.Name}' already exists");
            }
            existingCompanyGroup.Name = updateDto.Name;
        }

        if (updateDto.Description != null)
        {
            existingCompanyGroup.Description = updateDto.Description;
        }

        if (updateDto.IsActive.HasValue)
        {
            existingCompanyGroup.IsActive = updateDto.IsActive.Value;
        }

        await _context.SaveChangesAsync();
        _logger.LogInformation("Company group updated successfully with id: {Id}", existingCompanyGroup.Id);

        return ToDto(existingCompanyGroup);
    }

    public async Task DeleteAsync(Guid id)
    {
        _logger.LogInformation("Deleting company group with id: {Id}", id);

        var companyGroup = await _context.CompanyGroups
            .FirstOrDefaultAsync(cg => cg.Id == id);

        if (companyGroup == null)
        {
            throw new ArgumentException($"Company group not found with id: {id}");
        }

        _context.CompanyGroups.Remove(companyGroup);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Company group deleted successfully");
    }

    private static CompanyGroupDto ToDto(CompanyGroup companyGroup)
    {
        return new CompanyGroupDto
        {
            Id = companyGroup.Id,
            Name = companyGroup.Name,
            Description = companyGroup.Description,
            IsActive = companyGroup.IsActive,
            CreatedAt = companyGroup.CreatedAt,
            UpdatedAt = companyGroup.UpdatedAt
        };
    }
}